packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -446,13 +449,22 @@ class TextPainter {
         continue;
       }
       final TextBox box = boxes[0];
+
+      // If the upstream character is a newline, cursor is at start of next line
+      const int NEWLINE_CODE_UNIT = 10;
+      if (prevCodeUnit == NEWLINE_CODE_UNIT) {
+        return Offset(_emptyOffset.dx, box.bottom);
+      }
+
       final double caretEnd = box.end;
       final double dx = box.direction == TextDirection.rtl ? caretEnd - caretPrototype.width : caretEnd;
       return Offset(dx, box.top);
     }
     return null;
   }
 
+  // Get the Offset of the cursor (in pixels) based off the near edge of the
```

Comment:
May want to distinguish between logical vs physical pixels, as 'pixels' may be ambiguous. In this case I think everything should be logical pixels.